### PR TITLE
nixos/sssd: Add secrets handling

### PR DESCRIPTION
**Pull request by yayayayaka**
###### Motivation for this change
Currently, it is not possible to supply sensitive credentials like
`ldap_default_authtok` without writing them to the nix store. This
This commit introduces a new option `environmentFile` where those
credentials can be supplied via environment substitution.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rnhmjoj** [comment]
Related issue: https://github.com/NixOS/nixpkgs/issues/24288

File: nixos/modules/services/misc/sssd.nix [hunk 1]
```diff
@@ -38,6 +42,15 @@ in {
           For this to work, the <literal>ssh</literal> SSS service must be enabled in the sssd configuration.
         '';
       };
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          File containing environment variables to be passed to the sssd service,
```
Comment (rnhmjoj) [inline nixos/modules/services/misc/sssd.nix:49]: There's a good description format that's shared between a number of `environmentFile` options throughout NixOS:
- [services.hedgedoc.environmentFile](https://search.nixos.org/options?channel=21.05&from=0&size=50&sort=relevance&type=packages&query=services.hedgedoc.environmentFile)
- [services.prometheus.environmentFile](https://search.nixos.org/options?channel=21.05&show=services.prometheus.environmentFile&from=0&size=50&sort=relevance&type=packages&query=services.prometheus.environmentFile)
- [services.murmur.environmentFile](https://search.nixos.org/options?channel=21.05&show=services.murmur.environmentFile&from=0&size=50&sort=relevance&type=packages&query=services.murmur.environmentFile)
I'd advise to use it for this option too.

Comment (yayayayaka) [reply nixos/modules/services/misc/sssd.nix:49]: Thanks for your input! That looks indeed quite useful. I'm going to update the description later (possibly tonight).

And it is good to know that #24288 exists. I wasn't aware of that :)

**Comment by rnhmjoj** [comment]
Thanks. I can offer a few advices because I've just implemented something like this for another module, but keep in mind that I don't know or use sssd.

The first problem I see is that `envsubst` expand empty (or missing) variables, this may not what you want: in particular it can expand something that isn't a variable if the file contains a literal `$`. Is this a valid concern?

**Comment by aanderse** [comment]
After reading the man pages for `sssd.conf` I wonder if we're better off adding a structured attribute set (`settings`) to replace the `config` option and then telling users to put secrets in `/etc/sssd/conf.d`. The man pages state that the configuration will be merged so that sounds like a pretty solid plan to me...

I'd be willing to put a PR like that together.

**Comment by yayayayaka** [comment]
> The first problem I see is that envsubst expand empty (or missing) variables, this may not do what you want: in particular it can expand something that isn't a variable if the file contains a literal $. Is this a valid concern?

I am aware of that. As I am not a heavy user of sssd myself (only use it for the LDAP part), I don't think the `$` is used anywhere else in the config (but I'm not entirely sure).

I originally thought about converting the configuration to an attribute set, too as @aanderse suggested, but I thought it would better fit into a separate PR.

**Comment by aanderse** [comment]
> I originally thought about converting the configuration to an attribute set, too as @aanderse suggested, but I thought it would better fit into a separate PR.

My point is, though, that if you use an attribute set and combine that with `/etc/sssd/conf.d` files you don't need any substitutions at all - the module would work incredibly well with secrets.

**Comment by yayayayaka** [comment]
I personally don't need this PR to be merged anymore.

**Pull request closed without merging**